configure.ac

```diff
@@ -47,10 +43,37 @@ fi
 AC_SUBST([GUILE_TARGET])
 
 WARN_CFLAGS=-Wall
-AC_ARG_ENABLE([Werror], AC_HELP_STRING([--disable-Werror],[Don't stop the build on errors]),
+AC_ARG_ENABLE([Werror], AS_HELP_STRING([--disable-Werror],[Don't stop the build on errors]),
         [], WARN_CFLAGS="-Wall -Werror")
 AC_SUBST(WARN_CFLAGS)
 
+AC_ARG_WITH([libevent],
```

Comment:
Mmmm, I believe it's better to be more explicit and not enable things magically. I've always seen `--with-XXX` something the user needs to set explicitly. For example, when packaging a library, it's good to see `--with-libevent` when calling `./configure`. Someone who reads the packaging file will know what's going on. 

The way it is now, people that wants the default don't need to do anything. People that wants libevent they just need to set `--with-libevent`, in both cases no one requires to modify `configure.ac`.

I'll not mark this resolved for now and give it more thought.

Reply:
As a comment on autoconf culture, I see enable/disable as more about turning things on/off, and --with as more about giving the path to something.  However, that may be a hallucination that has shifted from something else that was true 25 years ago.   But, I think by default the config should first try to use (in an AC_TRY_LINK sort of way, roughly) the preferred backend, and then the next, sort of in the order that they would be preferred if they all worked.  For portability, I think it would be great if one could --disable-epoll on GNU/Linux and end up with libevent, and further that this work and be subject to regression tests.

Is the idea that if there is no epoll then the build fails, and the user has to read the config help and ask for it?  That seems strange, compared to checking mechanisms until one that works  is found.

Reply:
I see. It's true that `--with-libevent` it's actually not autoconf idiomatic, the `--enable-libevent` would be better. But since you both agree, I will follow your suggestions. I like the `--disable-epoll` but what if we called it `--disable-native` so we don't need to support `--disable-epoll`, `--disable-kqueue`?

Reply:
I am not clear on existing practice for finding one of multiple methods to do something.  However it seems pretty straightforward to just set up autoconf with enable for epoll then libevent, and then not worry that if multiple are found that there is no way to control priority.

Actually I wonder if libevent supports epoll, and if so, if there is any real downside to just "use libevent", period? It seems that libevent is the portable solution to the general problem, wrapping whatever exists.   On my netbsd-9 system, 2.1.8 is in the base system and 2.1.12 is in pkgsrc, and in pkgsrc it is "only" 2 MB and doesn't depend on anything.   I have a Debian machine on which I didn't intend to install that much and 2.1.7 is present.   I have a RPI3 running Raspberry Pi OS or Debian and it too has 2.1.7.   Is it normal on any system with a non-super-stripped installation to not have libevent already?  Do other programs use raw epoll and fall back?   Would fibers be slower?  Is there some reproducible tower argument (for the guix folks) against libevent?

Reply:
Alright. This is done. We now try `epoll`, if not found we try `libevent`. You can also disable `epoll` with `--disable-native` and that will then check for `libevent`. If `libevent` is not found it will error out saying not native event system support or libevent was found.

Reply:
I believe @wingo preferred native support instead of another dependecy. The reason is that he wants fibers to be included in Guile at some point.

Reply:
Marking this as resolved. Please, unresolve if I misunderstood and implemented something else.

Reply:
Well, you don't really have native support.  You have libevent on N-1 operating systems and native epoll on 1, as I understand it, so guile will need libevent on N-1 systems.   And surely it's actually there already.   But as long as it can work portably with libevent, it isn't going to cause me trouble.

Reply:
That's true. I just got the impression that Andy preferred `epoll`. Anyways, we can make the decision when we get there, for now we prefer native support. In terms of performance I don't think there's a huge difference, I believe epoll is slightly better  than libevent but I don't have the data to corroborate that (just from the timings you see during unit tests).

Reply:
I was thinking that all native supports `epoll`, `kqueue` or others all use `--disable-native`. Since we currently only have epoll, I'll rename it to `--disable-epoll`, that will certainly be less confusing.

Updated!

Reply:
TBC, I don't care about naming '--with-whatever' or '--enable-whatever'

> Is the idea that if there is no epoll then the build fails, and the user
> has to read the config help and ask for it? That seems strange,
> compared to checking mechanisms until one that works is found.

I don't know what you mean, the 'bail out if no epoll found' only applies to Linux systems, and Linux always has epoll (well, except for very old versions, and theoretically you could disable epoll in the kernel configuration, but meh).

Reply:
I meant that I am not really following the big picture.  If there is support for epoll, then it seems like that should be checked for first and used if present, and if not, then libevent should  be looked for (with pkg-config presumably) and if present used.  So there's "--disable-epoll" which forces off the epoll check so libevent will be used on Linux, which is great for CI, useful for answering the comparitive speed etc.   I think with some of the previous iterations I thought that configure.ac would only look for libevent if turned on explicitly, and that seemed odd -- but I hope that was just confusion on my part.   I think we've ended up in a pretty good place.

Reply:
Yes, initially we had to specify `--with-libevent`. Now we will try `epoll` (or whatever other native system we add in the future) and if it's not found we will check for `libevent`. If you are in a Linux system and want to work on `libevent` then you now just specify `--disable-epoll`. I believe that's what both of you suggested and I believe that it end up looking pretty good.